File: go/mechanisms/evm/service.go
package evm

import (
	"context"
	"fmt"
	"math/big"
	"strings"

	x402 "github.com/coinbase/x402/go"
)

// ExactEvmService implements the SchemeNetworkService interface for EVM exact payments (V2)
type ExactEvmService struct{}

// NewExactEvmService creates a new ExactEvmService
func NewExactEvmService() *ExactEvmService {
	return &ExactEvmService{}
}

// Scheme returns the scheme identifier
func (s *ExactEvmService) Scheme() string {
	return SchemeExact
}

// ParsePrice parses a price string and converts it to an asset amount (V2)
func (s *ExactEvmService) ParsePrice(price x402.Price, network x402.Network) (x402.AssetAmount, error) {
	// Convert price to string (Price is interface{})
	priceStr, ok := price.(string)
	if !ok {
		priceStr = fmt.Sprintf("%v", price)
	}
	networkStr := string(network)

	// Handle different price formats
	// Format 1: "$1.00" or "1.00 USD"
	// Format 2: "1000000" (already in smallest unit)
	// Format 3: "1.00" (decimal amount)

	// Remove common currency symbols and spaces
	priceStr = strings.TrimSpace(priceStr)
	priceStr = strings.TrimPrefix(priceStr, "$")
	priceStr = strings.TrimSuffix(priceStr, " USD")
	priceStr = strings.TrimSuffix(priceStr, " USDC")
	priceStr = strings.TrimSpace(priceStr)

	// Get network config to determine the asset
	config, err := GetNetworkConfig(networkStr)
	if err != nil {
		return x402.AssetAmount{}, err
	}

	// Try to parse as decimal first
	if strings.Contains(priceStr, ".") {
		// It's a decimal amount, convert to smallest unit
		amount, err := ParseAmount(priceStr, config.DefaultAsset.Decimals)
		if err != nil {
			return x402.AssetAmount{}, fmt.Errorf("failed to parse decimal price: %w", err)
		}

		return x402.AssetAmount{
			Asset:  config.DefaultAsset.Address,
			Amount: amount.String(),
		}, nil
	}

	// Try to parse as integer (already in smallest unit)
	amount, ok := new(big.Int).SetString(priceStr, 10)
	if !ok {
		return x402.AssetAmount{}, fmt.Errorf("invalid price format: %s", price)
	}

	// Check if this looks like a reasonable amount in smallest unit
	// (e.g., 1000000 for $1.00 USDC with 6 decimals)
	oneUnit := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(config.DefaultAsset.Decimals)), nil)
	if amount.Cmp(oneUnit) >= 0 {
		// Likely already in smallest unit
		return x402.AssetAmount{
			Asset:  config.DefaultAsset.Address,
			Amount: amount.String(),
		}, nil
	}

	// Small integer, treat as dollar amount
	amount.Mul(amount, oneUnit)

	return x402.AssetAmount{
		Asset:  config.DefaultAsset.Address,
		Amount: amount.String(),
	}, nil
}

// EnhancePaymentRequirements adds scheme-specific enhancements to payment requirements (V2)
func (s *ExactEvmService) EnhancePaymentRequirements(
	ctx context.Context,
	requirements x402.PaymentRequirements,
	supportedKind x402.SupportedKind,
	extensionKeys []string,
) (x402.PaymentRequirements, error) {
	// V2 specific: only handle version 2
	if supportedKind.X402Version != 2 {
		return requirements, fmt.Errorf("v2 only supports x402 version 2")
	}

	// Get network config
	networkStr := string(requirements.Network)
	config, err := GetNetworkConfig(networkStr)
	if err != nil {
		return requirements, err
	}

	// Get asset info
	var assetInfo *AssetInfo
	if requirements.Asset != "" {
		assetInfo, err = GetAssetInfo(networkStr, requirements.Asset)
		if err != nil {
			return requirements, err
		}
	} else {
		// Use default asset if not specified
		assetInfo = &config.DefaultAsset
		requirements.Asset = assetInfo.Address
	}

	// Ensure amount is in the correct format (smallest unit)
	if requirements.Amount != "" && strings.Contains(requirements.Amount, ".") {
		// Convert decimal to smallest unit
		amount, err := ParseAmount(requirements.Amount, assetInfo.Decimals)
		if err != nil {
			return requirements, fmt.Errorf("failed to parse amount: %w", err)
		}
		requirements.Amount = amount.String()
	}

	// Add EIP-3009 specific fields to Extra if not present
	if requirements.Extra == nil {
		requirements.Extra = make(map[string]interface{})
	}

	// Add token name and version for EIP-712 signing
	// ONLY add if not already present (client may have specified exact values)
	if _, ok := requirements.Extra["name"]; !ok {
		requirements.Extra["name"] = assetInfo.Name
	}
	if _, ok := requirements.Extra["version"]; !ok {
		requirements.Extra["version"] = assetInfo.Version
	}

	// Copy extensions from supportedKind if provided
	if supportedKind.Extra != nil {
		for _, key := range extensionKeys {
			if val, ok := supportedKind.Extra[key]; ok {
				requirements.Extra[key] = val
			}
		}
	}

	return requirements, nil
}

// GetDisplayAmount formats an amount for display
func (s *ExactEvmService) GetDisplayAmount(amount string, network string, asset string) (string, error) {
	// Get asset info
	assetInfo, err := GetAssetInfo(network, asset)
	if err != nil {
		return "", err
	}

	// Parse amount
	amountBig, ok := new(big.Int).SetString(amount, 10)
	if !ok {
		return "", fmt.Errorf("invalid amount: %s", amount)
	}

	// Format with decimals
	formatted := FormatAmount(amountBig, assetInfo.Decimals)

	// Add currency symbol
	return "$" + formatted + " USDC", nil
}

// ValidatePaymentRequirements validates that requirements are valid for this scheme
func (s *ExactEvmService) ValidatePaymentRequirements(requirements x402.PaymentRequirements) error {
	// Check network is supported
	networkStr := string(requirements.Network)
	if !IsValidNetwork(networkStr) {
		return fmt.Errorf("unsupported network: %s", requirements.Network)
	}

	// Check PayTo is a valid address
	if !IsValidAddress(requirements.PayTo) {
		return fmt.Errorf("invalid PayTo address: %s", requirements.PayTo)
	}

	// Check amount is valid
	if requirements.Amount == "" {
		return fmt.Errorf("amount is required")
	}

	amount, ok := new(big.Int).SetString(requirements.Amount, 10)
	if !ok || amount.Sign() <= 0 {
		return fmt.Errorf("invalid amount: %s", requirements.Amount)
	}

	// Check asset is valid if specified
	if requirements.Asset != "" && !IsValidAddress(requirements.Asset) {
		// Try to look it up as a symbol
		_, err := GetAssetInfo(networkStr, requirements.Asset)
		if err != nil {
			return fmt.Errorf("invalid asset: %s", requirements.Asset)
		}
	}

	return nil
}

// ConvertToTokenAmount converts a decimal amount to token smallest unit
func (s *ExactEvmService) ConvertToTokenAmount(decimalAmount string, network string) (string, error) {
	config, err := GetNetworkConfig(network)
	if err != nil {
		return "", err
	}

	amount, err := ParseAmount(decimalAmount, config.DefaultAsset.Decimals)
	if err != nil {
		return "", err
	}

	return amount.String(), nil
}

// ConvertFromTokenAmount converts from token smallest unit to decimal
func (s *ExactEvmService) ConvertFromTokenAmount(tokenAmount string, network string) (string, error) {
	config, err := GetNetworkConfig(network)
	if err != nil {
		return "", err
	}

	amount, ok := new(big.Int).SetString(tokenAmount, 10)
	if !ok {
		return "", fmt.Errorf("invalid token amount: %s", tokenAmount)
	}

	return FormatAmount(amount, config.DefaultAsset.Decimals), nil
}

// GetSupportedNetworks returns the list of supported networks
func (s *ExactEvmService) GetSupportedNetworks() []string {
	networks := make([]string, 0, len(NetworkConfigs))
	for network := range NetworkConfigs {
		networks = append(networks, network)
	}
	return networks
}

// GetSupportedAssets returns the list of supported assets for a network
func (s *ExactEvmService) GetSupportedAssets(network string) ([]string, error) {
	config, err := GetNetworkConfig(network)
	if err != nil {
		return nil, err
	}

	assets := make([]string, 0, len(config.SupportedAssets))
	for symbol := range config.SupportedAssets {
		assets = append(assets, symbol)
	}

	// Also add the addresses
	for _, asset := range config.SupportedAssets {
		assets = append(assets, asset.Address)
	}

	return assets, nil
}

File: typescript/packages/mechanisms/evm/src/exact/service.ts
import {
  AssetAmount,
  Network,
  PaymentRequirements,
  Price,
  SchemeNetworkService,
} from "@x402/core/types";

/**
 * EVM service implementation for the Exact payment scheme.
 */
export class ExactEvmService implements SchemeNetworkService {
  readonly scheme = "exact";

  /**
   * Parses a price into an asset amount.
   *
   * @param price - The price to parse
   * @param network - The network to use
   * @returns The parsed asset amount
   */
  parsePrice(price: Price, network: Network): AssetAmount {
    // Handle pre-parsed price object
    if (typeof price === "object" && price !== null && "amount" in price) {
      if (!price.asset) {
        throw new Error(`Asset address must be specified for price object on network ${network}`);
      }
      return {
        amount: price.amount,
        asset: price.asset,
        extra: price.extra || {},
      };
    }

    // Parse string prices like "$0.10" or "0.10 USDC"
    if (typeof price === "string") {
      // Remove $ sign if present
      const cleanPrice = price.replace(/^\$/, "").trim();

      // Check if it contains a currency/asset identifier
      const parts = cleanPrice.split(/\s+/);
      if (parts.length === 2) {
        // Format: "0.10 USDC"
        const amount = this.convertToTokenAmount(parts[0], network);
        const assetInfo = this.getAssetInfo(parts[1], network);
        return {
          amount,
          asset: assetInfo.address,
          extra: {
            name: assetInfo.name,
            version: assetInfo.version,
          },
        };
      } else if (cleanPrice.match(/^\d+(\.\d+)?$/)) {
        // Simple number format like "0.10" - assume USD/USDC
        const amount = this.convertToTokenAmount(cleanPrice, network);
        const assetInfo = this.getDefaultAsset(network);
        return {
          amount,
          asset: assetInfo.address,
          extra: {
            name: assetInfo.name,
            version: assetInfo.version,
          },
        };
      } else {
        throw new Error(
          `Invalid price format: ${price}. Must specify currency (e.g., "0.10 USDC") or use simple number format.`,
        );
      }
    }

    // Handle number input - assume USD/USDC
    if (typeof price === "number") {
      const amount = this.convertToTokenAmount(price.toString(), network);
      const assetInfo = this.getDefaultAsset(network);
      return {
        amount,
        asset: assetInfo.address,
        extra: {
          name: assetInfo.name,
          version: assetInfo.version,
        },
      };
    }

    throw new Error(`Invalid price format: ${price}`);
  }

  /**
   * Build payment requirements for this scheme/network combination
   *
   * @param paymentRequirements - The base payment requirements
   * @param supportedKind - The supported kind from facilitator (unused)
   * @param supportedKind.x402Version - The x402 version
   * @param supportedKind.scheme - The logical payment scheme
   * @param supportedKind.network - The network identifier in CAIP-2 format
   * @param supportedKind.extra - Optional extra metadata regarding scheme/network implementation details
   * @param extensionKeys - Extension keys supported by the facilitator (unused)
   * @returns Payment requirements ready to be sent to clients
   */
  enhancePaymentRequirements(
    paymentRequirements: PaymentRequirements,
    supportedKind: {
      x402Version: number;
      scheme: string;
      network: Network;
      extra?: Record<string, unknown>;
    },
    extensionKeys: string[],
  ): Promise<PaymentRequirements> {
    // Mark unused parameters to satisfy linter
    void supportedKind;
    void extensionKeys;
    return Promise.resolve(paymentRequirements);
  }

  /**
   * Convert decimal amount to token units (e.g., 0.10 -> 100000 for 6-decimal USDC)
   *
   * @param decimalAmount - The decimal amount to convert
   * @param network - The network to use
   * @returns The token amount as a string
   */
  private convertToTokenAmount(decimalAmount: string, network: Network): string {
    const decimals = this.getAssetDecimals(network);
    const amount = parseFloat(decimalAmount);
    if (isNaN(amount)) {
      throw new Error(`Invalid amount: ${decimalAmount}`);
    }
    // Convert to smallest unit (e.g., for USDC with 6 decimals: 0.10 * 10^6 = 100000)
    const tokenAmount = Math.floor(amount * Math.pow(10, decimals));
    return tokenAmount.toString();
  }

  /**
   * Get the default asset info for a network (typically USDC)
   *
   * @param network - The network to get asset info for
   * @returns The asset information including address, name, and version
   */
  private getDefaultAsset(network: Network): { address: string; name: string; version: string } {
    // Map of network to USDC info including EIP-712 domain parameters
    const usdcInfo: Record<string, { address: string; name: string; version: string }> = {
      "eip155:8453": {
        address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
        name: "USD Coin",
        version: "2",
      }, // Base mainnet USDC
      "eip155:84532": {
        address: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
        name: "USDC",
        version: "2",
      }, // Base Sepolia USDC
      "eip155:1": {
        address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        name: "USD Coin",
        version: "2",
      }, // Ethereum mainnet USDC
      "eip155:11155111": {
        address: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238",
        name: "USDC",
        version: "2",
      }, // Sepolia USDC
    };

    const assetInfo = usdcInfo[network];
    if (!assetInfo) {
      throw new Error(`No default asset configured for network ${network}`);
    }

    return assetInfo;
  }

  /**
   * Get asset info for a given symbol on a network
   *
   * @param symbol - The asset symbol
   * @param network - The network to use
   * @returns The asset information including address, name, and version
   */
  private getAssetInfo(
    symbol: string,
    network: Network,
  ): { address: string; name: string; version: string } {
    const upperSymbol = symbol.toUpperCase();

    // For now, only support USDC
    if (upperSymbol === "USDC" || upperSymbol === "USD") {
      return this.getDefaultAsset(network);
    }

    // Could extend to support other tokens
    throw new Error(`Unsupported asset: ${symbol} on network ${network}`);
  }

  /**
   * Get the number of decimals for the asset
   *
   * @param _ - The network to use (unused)
   * @returns The number of decimals for the asset
   */
  private getAssetDecimals(_: Network): number {
    // USDC has 6 decimals on all EVM chains
    return 6;
  }
}
